File: sortingAlgo/mergeSort/mergeSort.py
# Testcase
arr = [63, 1, 103, 3, 8, 74, 76, 37, 71, 66]
l = len(arr)

def mergeSort(arr):
	l = len(arr)
	mid = l//2

	if l <= 1:
		return arr

	left = arr[:mid]
	right = arr[mid:]			#_______
								#	  	|
	left = mergeSort(left)		#       |____ Understand
	right = mergeSort(right)	#		|     this properly!
								#_______|
								#
	return merge(left, right)


def merge(arr1, arr2):
	arr = []
	l1 = len(arr1)
	l2 = len(arr2)

	while((l1 != 0) and (l2 != 0)):
		if arr1[0] > arr2[0]:
			arr.append(arr2[0])
			del arr2[0]
			l2 -= 1
		elif arr1[0] < arr2[0]:
			arr.append(arr1[0])
			del arr1[0]
			l1 -= 1
		else:
			arr.append(arr1[0])
			arr.append(arr2[0])
			del arr2[0]
			del arr1[0]
			l1 -= 1
			l2 -= 1

	if l1 == 0:
		arr.extend(arr2)
	else:
		arr.extend(arr1)

	return arr

print(mergeSort(arr))

File: sortingAlgo/mergeSort/mergeSort.lua
local function mergeSort(t)
    local n = #t
    if n < 2 then
        return t
    end
    local mid = math.floor(n / 2)
    local left = {}
    local right = {}
    for i = 1, mid do
        left[i] = t[i]
    end
    for i = mid + 1, n do
        right[i - mid] = t[i]
    end
    left = mergeSort(left)
    right = mergeSort(right)
    local i = 1
    local j = 1
    local k = 1
    while i <= #left and j <= #right do
        if left[i] < right[j] then
            t[k] = left[i]
            i = i + 1
        else
            t[k] = right[j]
            j = j + 1
        end
        k = k + 1
    end
    while i <= #left do
        t[k] = left[i]
        i = i + 1
        k = k + 1
    end
    while j <= #right do
        t[k] = right[j]
        j = j + 1
        k = k + 1
    end
    return t
end

local t = { 3, 2, 1, 4, 5, 6, 7, 8, 9, 10 }
mergeSort(t)
print(table.concat(t, ", "))